python/src/pymimir/bindings.cpp

```diff
@@ -341,727 +330,682 @@ void init_pymimir(py::module_& m)
+                    py::keep_alive<0, 1>())
+            .def(
+                    "get_numeric_fluents",
+                    [](const ProblemImpl& self) { return NumericFluentList(self.get_numeric_fluents()); },
+                    py::keep_alive<0, 1>())
+            .def("get_optimization_metric", &ProblemImpl::get_optimization_metric, py::return_value_policy::reference_internal)
+            .def(
+                    "get_static_goal_condition",
+                    [](const ProblemImpl& self) { return GroundLiteralList<Static>(self.get_goal_condition<Static>()); },
+                    py::keep_alive<0, 1>())
+            .def(
+                    "get_fluent_goal_condition",
+                    [](const ProblemImpl& self) { return GroundLiteralList<Fluent>(self.get_goal_condition<Fluent>()); },
+                    py::keep_alive<0, 1>())
+            .def(
+                    "get_derived_goal_condition",
+                    [](const ProblemImpl& self) { return GroundLiteralList<Derived>(self.get_goal_condition<Derived>()); },
+                    py::keep_alive<0, 1>());
     static_assert(!py::detail::vector_needs_copy<ProblemList>::value);  // Ensure return by reference + keep alive
-    py::bind_vector<ProblemList>(m, "ProblemList");
+    list_class = py::bind_vector<ProblemList>(m, "ProblemList");
+    def_opaque_vector_repr<ProblemList>(list_class, "ProblemList");
 
     py::class_<PDDLFactories, std::shared_ptr<PDDLFactories>>(m, "PDDLFactories")  //
-        .def("get_static_ground_atom", &PDDLFactories::get_ground_atom<Static>, py::return_value_policy::reference_internal)
-        .def("get_fluent_ground_atom", &PDDLFactories::get_ground_atom<Fluent>, py::return_value_policy::reference_internal)
-        .def("get_derived_ground_atom", &PDDLFactories::get_ground_atom<Derived>, py::return_value_policy::reference_internal)
-        .def("get_static_ground_atoms_from_indices",
-             py::overload_cast<const std::vector<size_t>&>(&PDDLFactories::get_ground_atoms_from_indices<Static, std::vector<size_t>>, py::const_),
-             py::keep_alive<0, 1>())
-        .def("get_fluent_ground_atoms_from_indices",
-             py::overload_cast<const std::vector<size_t>&>(&PDDLFactories::get_ground_atoms_from_indices<Fluent, std::vector<size_t>>, py::const_),
-             py::keep_alive<0, 1>())
-        .def("get_derived_ground_atoms_from_indices",
-             py::overload_cast<const std::vector<size_t>&>(&PDDLFactories::get_ground_atoms_from_indices<Derived, std::vector<size_t>>, py::const_),
-             py::keep_alive<0, 1>())
-        .def("get_object", &PDDLFactories::get_object, py::return_value_policy::reference_internal);
+            .def("get_ground_atoms",
+                 [](py::object py_factory)  // we need an object handle to keep the factory alive for each atom in the list
+                 {
+                     const auto& factory = py::cast<const PDDLFactories&>(py_factory);
```

Comment:
Indeed, you can do:
```cpp
for (const auto& atom : factory.get_factory<GroundAtomFactory<Static>>()) { 
    // ...
}

```

Reply:
Good to know! Using the factory also gets rid of the unnecessary intermediate vector allocation